core/drivers/pl022_spi.c

```diff
@@ -0,0 +1,514 @@
+			(j < *num_rxpkts))
+			/* rx 1 packet */
+			rdat[j++] = read8(pd->base + SSPDR);
+	} while ((read8(pd->base + SSPSR) & SSPSR_BSY) && (j < *num_rxpkts));
+
+	*num_rxpkts = j;
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
+}
+
+static void pl022_txrx16(struct spi_chip *chip, uint16_t *wdat,
+	uint16_t *rdat, size_t num_txpkts, size_t *num_rxpkts)
+{
+	size_t i = 0;
+	size_t j = 0;
+	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
+
+	while (i < num_txpkts) {
+		if (read8(pd->base + SSPSR) & SSPSR_TNF)
+			/* tx 1 packet */
+			write16(wdat[i++], pd->base + SSPDR);
+	}
+
+	do {
+		while ((read8(pd->base + SSPSR) & SSPSR_RNE)
+			&& (j < *num_rxpkts))
+			/* rx 1 packet */
+			rdat[j++] = read16(pd->base + SSPDR);
+	} while ((read8(pd->base + SSPSR) & SSPSR_BSY) && (j < *num_rxpkts));
+
+	*num_rxpkts = j;
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
+}
+
+static void pl022_tx8(struct spi_chip *chip, uint8_t *wdat,
+	size_t num_txpkts)
+{
+	size_t i = 0;
+	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
+
+	while (i < num_txpkts) {
+		if (read8(pd->base + SSPSR) & SSPSR_TNF)
+			/* tx 1 packet */
+			write8(wdat[i++], pd->base + SSPDR);
+	}
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
+}
+
+static void pl022_tx16(struct spi_chip *chip, uint16_t *wdat,
+	size_t num_txpkts)
+{
+	size_t i = 0;
+	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
+
+	while (i < num_txpkts) {
+		if (read8(pd->base + SSPSR) & SSPSR_TNF)
+			/* tx 1 packet */
+			write16(wdat[i++], pd->base + SSPDR);
+	}
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
+}
+
+static void pl022_rx8(struct spi_chip *chip, uint8_t *rdat,
+	size_t *num_rxpkts)
+{
+	size_t j = 0;
+	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
+
+	do {
+		while ((read8(pd->base + SSPSR) & SSPSR_RNE) &&
+			(j < *num_rxpkts))
+			/* rx 1 packet */
+			rdat[j++] = read8(pd->base + SSPDR);
+	} while ((read8(pd->base + SSPSR) & SSPSR_BSY) && (j < *num_rxpkts));
+
+	*num_rxpkts = j;
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
+}
+
+static void pl022_rx16(struct spi_chip *chip, uint16_t *rdat,
+	size_t *num_rxpkts)
+{
+	size_t j = 0;
+	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
+
+	do {
+		while ((read8(pd->base + SSPSR) & SSPSR_RNE) &&
+			(j < *num_rxpkts))
+			/* rx 1 packet */
+			rdat[j++] = read16(pd->base + SSPDR);
+	} while ((read8(pd->base + SSPSR) & SSPSR_BSY) && (j < *num_rxpkts));
+
+	*num_rxpkts = j;
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
+}
+
+static void pl022_print_peri_id(struct pl022_data *pd __maybe_unused)
+{
+	DMSG("Expected: 0x 22 10 #4 0");
```

Comment:
`#` is supposed to denote an unknown number. Based on the pl022 spec, this is the only part of the peripheral ID that is not a fixed constant value. It returns the peripheral revision. Should I maybe change it to `22 10 [revision]4 0`?